dev/planning-example.py

```diff
@@ -0,0 +1,292 @@
+If no parameter is given, the current date/time is assumed.
+"""
+
+# sunset
+obs.next_sunset(date=day)
+obs.previous_sunset(date=day)
+
+# sunrise
+obs.next_sunrise(date=day)
+obs.previous_sunrise(date=day)
+
+# moon rise
+obs.next_moon_rise(date=day)
+obs.previous_moon_rise(date=day)
+
+# moon set
+obs.next_moon_set(date=day)
+obs.previous_moon_set(date=day)
+
+# The above functions can be called with an `angle` keyword argument to specify
+# a particular horizon angle for rising or setting, or can be called with
+# convenience functions for particular morning/evening twilight:
+
+obs.next_sunset(date=day, angle=18*u.degree) # (i.e., astronomical twilight)
+
+# evening (astronomical) twilight
+obs.evening_astronomical(date=day)
+
+# evening (nautical) twilight
+obs.evening_nautical(date=day)
+
+# evening (civil) twilight
+obs.evening_civil(date=day)
+
+# morning (nautical) twilight
+obs.morning_nautical(date=day)
+
+# morning (civil) twilight
+obs.morning_civil(date=day)
+
+# morning (astronomical) twilight
+obs.morning_astronomical(date=day)
+
+# what is the moon illumination?
+# returns a float, which is percentage of the moon illuminated
+obs.moon_illumination(date=t1)
+
+# what is the moon altitude and azimuth?
+# returns an altaz coordinate
+obs.moon_position(date=t1)
+
+# Other sun-related convenience functions:
+obs.noon(date=day)
+obs.midnight(date=day)
+
+# ==============
+# Target objects
+# ==============
+'''
+A target object defines a single observation target, with coordinates
+and an optional name.
+'''
+from astroplan import FixedTarget, airmass_plot
+
+# Define a target.
+t1 = FixedTarget(name='Polaris', ra='02:31:49.09', dec='+89:15:50.8')
+
+# initializing from astropy entities directly
+from astropy.coordinates import SkyCoord
+
+t1 = FixedTarget(name='Polaris',
+               coord=SkyCoord('02h31m49.09s', '+89d15m50.8s', frame='icrs'))
+
+# Leaves scope for NonSiderealTarget, etc.
+
+# Convenience methods for looking up/constructing targets by name via
+# astroquery?
+
+# Plot the airmass for the specified target:
+airmass_plot(t1, observer=obs, datetime=day)
+
+# ================================
+# Condition objects, observability
+# ================================
+
+"""
+Q: Can I observe this target on the night of May 1, 2015 between 18:30
+and 05:30 HST for 30 minutes somewhere between 15 degrees and 89 degrees
+altitude?
+"""
+
+# first we define the conditions
+from astroplan import Constraints, TimeWindow, AltitudeRange, AboveAirmass
+
+# times can be passed in as strings (interpreted as for get_date()) or
+# as astropy Time or datetime objects
+cons = [TimeWindow("2015-05-01 18:30", "2015-05-02 05:30"),
+        AltitudeRange(15.0*u.deg, 89.0*u.deg), AboveAirmass(1.2)]
+# define a target
+tgt = FixedTarget(name='S5', ra='14:20:00.00', dec='48:00:00.00')
+
+# returns an object with information about the observability with these
+# constraints
+info = obs.Constraints(tgt, cons)
```

Comment:
#### On Constraints class

I've been involved in a few conversations about improving the name of the constraints class in different emails and Hangouts, and wanted to put down some notes where all @astroplanners/astroplanners-contributors could comment one last time before I make a third iteration of the API.

The guiding example we last touched on was something like this: 

```python
constraint_list = [TimeWindow("2015-05-01 18:30", "2015-05-02 05:30"),
                   AltitudeWindow(15.0*u.deg, 89.0*u.deg), AboveAirmass(1.2)]
constraints = Constraints(constraint_list)
info = constraints(obs, tgt)
```

My most fundamental question here is – does creating an instance of `Constraints` return a function, or an object? It looks like this first line is creating an instance of the `Constraints` class, which returns a function `cons`. Then `cons` produces an object `result` with attributes [specified in the API](https://github.com/astroplanners/astroplan/pull/4/files#diff-d1475973061987d43ea4fea64513b64aR227).

##### My suggestion

I'm not sure how different what I'm about to suggest may be different from above or what everyone else is imagining, but here's a strategy that makes sense to me, and I'm interested in your comments on it:
* We define a class called `Constraint` which contains the information and machinery needed to determine when targets are/aren't visible
* We define subclasses of `Constraint` such as `TimeWindow`, `AltitudeWindow ` which make it easy to construct some common constraints
* We create another class called e.g. `Observability` which gets initialized with:
    * a `Constraint`, or a list of `Constraint`s
    * an instance of `Observer`
    * an instance of `FixedTarget`
* The `Observability` object has [attributes specified earlier](https://github.com/astroplanners/astroplan/pull/4/files#diff-d1475973061987d43ea4fea64513b64aR227).

This would look something like this: 

```python
from astroplan import TimeWindow, AltitudeWindow, AboveAirmass, Observability
constraint_list = [TimeWindow("2015-05-01 18:30", "2015-05-02 05:30"),
                   AltitudeWindow(15.0*u.deg, 89.0*u.deg), AboveAirmass(1.2)]
result = Observability(constraint_list, obs, tgt)
print(result.observable) # is target observable? [boolean]
```

Pros, cons? Am I making any sense?